docs/Technical-Documentation/tech-memos/parsing-refactor-plan.md

```diff
@@ -0,0 +1,719 @@
+  - Which tables are touched and in what order
+  - How ReparseMeta / ReparseFileMeta are created and updated
+- Add any missing indexes or small performance improvements that are clearly safe (see separate performance tickets).
+
+
+### Phase 2 – Introduce `ParsingService` without changing behavior
+
+- Extract the body of `parse_data_file` into `ParsingService.parse_data_file`, keeping behavior identical.
+- The Celery task delegates to the service, but inputs/outputs remain unchanged.
+- Add tests that assert the service produces the same side effects as the existing Celery task for a small fixture.
+
+
+### Phase 3 – Wire reparsing through `ParsingService`
+
+- Refactor `search_indexes/reparse.py` and `search_indexes/utils.py` so that they:
+  - Only perform backup + selection + deletion + Celey scheduling
+  - Never run parsing logic directly
+- Ensure that `ReparseMeta` / `ReparseFileMeta` updates are driven by `ReparseService` and not by ad-hoc logic outside the service.
+
+
+### Phase 4 – Clean up and harden
+
+- Remove now-dead code paths or duplicated logic.
+- Add better failure handling and idempotency guarantees for reparse runs:
+  - Safe retry behavior if a Celery worker crashes mid-run
+  - Clear reporting of which files succeeded/failed in `ReparseMeta`
+- Add regression tests for:
+  - Incremental reparsing (subset of files, STT-specific)
+  - Large batches (e.g., dozens/hundreds of files)
+
+
+## 6. Expected Benefits
+
+1. **Reduced duplication**  
+   Single source of truth for parsing logic, regardless of whether it is an initial parse or a reparse.
+
+2. **Easier reasoning and debugging**  
+   Parsing behavior is concentrated in `ParsingService`; reparse orchestration just chooses “what” to parse.
+
+3. **Better testability**  
+   Service-oriented design enables targeted unit tests instead of having to go through Celery + management commands for every behavior change.
+
+4. **Improved performance tuning**  
+   Backup, deletion, and parsing steps are separated more clearly, making it easier to profile and optimize the heaviest operations.
+
+5. **Lower operational risk**  
+   A more explicit lifecycle for ReparseMeta / ReparseFileMeta, with a single place where their states are updated, makes it easier to detect and recover from partial failures.
+
+
+## 7. Next Steps
+
+1. Create tickets for:
+   - Introducing `ParsingService` and refactoring the Celery task to use it
+   - Adding `ReparseService` and refactoring reparse orchestration to depend on it (while delegating per-file parsing to `ParsingService`)
+   - Adding tests and observability around the unified parsing pipeline
+2. Implement Phase 2 first (service extraction) with strict “no behavior change” to build confidence.
+3. Once stable in staging, implement the reparse refactor and validate on a controlled subset of files.
+
+
+
+# Parsing & Reparsing Refactor – Ticket Breakdown
+
+---
+
+## Ticket 0 – Implement submission state machine before parser refactor
```

Comment:
Please take a look into these suggested tickets and let me know what you think? size-scope-etc